libraries/AP_BattMonitor/AP_BattMonitor_Backend.cpp

```diff
@@ -214,4 +217,47 @@ void AP_BattMonitor_Backend::check_failsafe_types(bool &low_voltage, bool &low_c
     } else {
         low_capacity = false;
     }
+
+    // check battery cell voltages
+    if (has_cell_voltages()) {
+
+        uint16_t cell_diff_mv = 0.0f;
+        uint16_t cell_lowest_mv = _state.cell_voltages.cells[0];  // initialize with the first cell's voltage
+        uint16_t cell_highest_mv = _state.cell_voltages.cells[0]; // initialize with the first cell's voltage
+
+        for (uint8_t i = 0; i < ARRAY_SIZE(_state.cell_voltages.cells); i++) {
+            
+            // bail out if cell voltage array element is UINT16_MAX since that means we've reached elements
+            if (_state.cell_voltages.cells[i] == UINT16_MAX) {
+                break;
+            }
+            
+            // look for the lowest & highest cell voltage
+            cell_lowest_mv = MIN(cell_lowest_mv, _state.cell_voltages.cells[i]);
+            cell_highest_mv = MAX(cell_highest_mv, _state.cell_voltages.cells[i]);
+        }
+
+        // calc mv difference between the lowest & highest cells
+        cell_diff_mv = cell_highest_mv - cell_lowest_mv;
+
+        // critical cell voltage
+        if (is_positive(_params._crt_cell_voltage) && ((float)(cell_lowest_mv * 1e-3) < _params._crt_cell_voltage)) {
+            critical_cell_voltage = true;
+        } else {
+            critical_cell_voltage = false;
+        }
+
+        // low cell voltage or cell highest to lowest diff
+        if ((is_positive(_params._low_cell_voltage) && ((float)(cell_lowest_mv * 1e-3) < _params._low_cell_voltage)) || 
+           (is_positive(_params._cell_voltage_diff) && ((float)(cell_diff_mv * 1e-3) > _params._cell_voltage_diff))) {
```

Comment:
Add an extra boolean argument for the critical version. (Which allows you to have a proper error string for that).

Reply:
Long term (but out of scope/not needed for this PR), I should really just make that a struct that we fill in, the list has gotten unwieldy long)

Reply:
Good point on the critical bool.  Done.